tenseal/utils/matrix_ops.h

```diff
@@ -97,6 +99,79 @@ Ciphertext diagonal_ct_vector_matmul(shared_ptr<TenSEALContext> tenseal_context,
     return result;
 }
 
+template <typename T, class Encoder>
+Ciphertext diagonal_ct_vector_matmul_parallel(
+    shared_ptr<TenSEALContext> tenseal_context, Ciphertext& vec,
+    size_t vector_size, const vector<vector<T>>& matrix, uint n_threads = 0) {
+    // matrix is organized by rows
+    // _check_matrix(matrix, this->size())
+    size_t n_rows = matrix.size();
+    size_t n_cols = matrix[0].size();
+
+    if (vector_size != matrix.size()) {
+        throw invalid_argument("matrix shape doesn't match with vector size");
+    }
+
+    mutex result_mutex;
+    Ciphertext result;
+    // result should have the same scale and modulus as vec * pt_diag (ct)
+    tenseal_context->encryptor->encrypt_zero(vec.parms_id(), result);
+    result.scale() = vec.scale() * tenseal_context->global_scale();
+
+    mutex i_mutex;
+    size_t i = 0;
+    auto thread_func = [&]() {
+        while (true) {
+            // take next i
+            size_t local_i;
+            i_mutex.lock();
+            if (i == n_rows) {
+                i_mutex.unlock();
+                break;
+            }
+            local_i = i++;
+            i_mutex.unlock();
+
+            Ciphertext ct;
+            Plaintext pt_diag;
+            vector<T> diag;
+
+            diag = get_diagonal(matrix, -local_i,
+                                tenseal_context->slot_count<Encoder>());
+            replicate_vector(diag, tenseal_context->slot_count<Encoder>());
+
+            rotate(diag.begin(), diag.begin() + diag.size() - local_i,
+                   diag.end());
+
+            tenseal_context->encode<Encoder>(diag, pt_diag);
+
+            if (vec.parms_id() != pt_diag.parms_id()) {
+                set_to_same_mod(tenseal_context, vec, pt_diag);
+            }
+            tenseal_context->evaluator->multiply_plain(vec, pt_diag, ct);
+
+            tenseal_context->evaluator->rotate_vector_inplace(
+                ct, local_i, *tenseal_context->galois_keys());
+
+            // accumulate results
+            result_mutex.lock();
+            tenseal_context->evaluator->add_inplace(result, ct);
+            result_mutex.unlock();
+        }
+    };
+
+    // if not specified (n_threads set to 0), detect automatically
+    if (n_threads == 0) n_threads = get_concurrency();
+    vector<thread> threads;
+    threads.reserve(n_threads);
+    // start the threads
+    for (int i = 0; i < n_threads; i++) threads.push_back(thread(thread_func));
```

Comment:
It can lead to performance issues to create the threads on every multiplication.

I think the threadpool pattern is much better suited here: at start you create N thread workers, and you pass the lambdas, with the global result in their context. You just have to wait for your workers to finish your specific task.

This can be achieved with a condition variable per task https://en.cppreference.com/w/cpp/thread/condition_variable

Reply:
Will investigate this

Reply:
It's also worth considering that some operations may take seconds, how much speed-up will thread pooling lead in that case?

Reply:
If there's a burst of threads, it will bring a major improvement. 
It's not usually ok to create threads on user demand, in can lead to denial of service if we keep requesting threads from the kernel.